source/fundamentals/gridfs.txt

```diff
@@ -4,7 +4,371 @@ GridFS
+to a GridFS bucket. It uses an ``opts`` parameter to set file metadata:
+
+.. io-code-block::
+   :copyable: true
+
+   .. input::
+      :language: go
+
+      fileContent, err := os.Open("path/to/file.txt")
+      uploadOpts := options.GridFSUpload().SetMetadata(bson.D{{"metadata tag", "first"}})
+
+      objectID, err := bucket.UploadFromStream("file.txt", io.Reader(fileContent),
+         uploadOpts)
+      if err != nil {
+         panic(err)
+      }
+
+      fmt.Printf("New file uploaded with ID %s", objectID)
+
+   .. output::
+      :language: none
+      :visible: false
+
+      New file uploaded with ID 62e005408bc04f3816b8bcd2
+
+.. note::
+
+   The driver uniquely generates each object ID number. The ID number outputted will
+   resemble the sample output but varies with each file and user.
+
+Upload with an Output Stream
+````````````````````````````
+
+You can use an output stream to upload a file to a GridFS bucket. To upload a file,
+use the ``OpenUploadStream()`` method with the following parameters:
+
+- Your file name
+- An optional ``opts`` parameter to modify the behavior of ``OpenUploadStream()``
+
+The following code example opens an upload stream on a GridFS bucket and sets the number
+of bytes in each chunk with an ``opts`` parameter. Then, it calls the ``Write()`` method
+on the content of ``"file.txt"`` to write its content to the stream:
+
+.. io-code-block::
+   :copyable: true
+
+   .. input::
+      :language: go
+
+      fileContent, err := os.Open("path/to/file.txt")
+      uploadOpts := options.GridFSUpload().SetChunkSizeBytes(200000)
+      uploadStream, err := bucket.OpenUploadStream("file.txt", uploadOpts)
+
+      if bytes, err = uploadStream.Write(fileContent); err != nil {
+         panic(err)
+      }
+
+      fmt.Printf("New file uploaded with %d bytes written", bytes)
+
+   .. output::
+      :language: none
+      :visible: false
+
+      New file uploaded with 1000 bytes written
+
+.. _golang-retrieve-info:
+
+Retrieve File Information
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can retrieve file metadata stored in the ``files`` collection of the GridFS
+bucket. Each document in the ``files`` collection contains the following information:
+
+- The file ID
+- The file length
+- The maximum chunk size
+- The upload date and time
+- The file name
+- A ``metadata`` document in which you can store any other information
+
+To retrieve file data, call the ``Find()`` method on a ``GridFSBucket``
+instance. You can pass a query filter as an argument to ``Find()`` to match
+only certain file documents.
+
+.. note::
+
+   The ``Find()`` method requires a query filter as a parameter. To match all 
+   documents in the ``files`` collection, pass an empty query filter to ``Find()``.
+
+The following example retrieves the file name and length of documents in the ``files``
+collection. It passes a filter argument to the ``Find()`` method to match documents with
+``length`` values greater than ``1500``:
```

Comment:
Suggestion: This could be simplified to explain what the code is doing without going too much into implementation, since that is done in the code itself:

```suggestion
The following example retrieves the name and length of all documents with a ``length`` value greater than 1500 in the ``files`` collection.
```

Reply:
Changed this